challenges/2021-11-16-data-grouping/solutions/javascript/justinmemphis/justinmemphis.js

```diff
@@ -0,0 +1,156 @@
+
+```
+HighSchool Year  Mean     Median
+Freshman         20       20
+Sophomore        26.5     26.5
+Junior           26       26
+Senior           28       28
+```
+
+###
+# outputs of:
+
+#return the number of scores greater than 25.
+3
+
+#return the number of sophomore scores greater than 25 and less than 30.
+0
+
+#return the mean of only the sophomore and junior scores.
+26.25
+
+## Tackling This Challenge
+
+1. Make sure you've got the required software on your machine: python 3.7+, pandas >= 1.0
+2. If you haven't already, fork the CodingDojo repository ([INSTRUCTIONS](https://docs.github.com/en/github/getting-started-with-github/fork-a-repo)).
+4. The challenge will be live-coded in our weekly Tusday meetup in the `solution.py` file.
+6. The results of the live coding demo will be added to the GitHub repository as a mob solution. Please feel free to tackle this challenge in your own way/language and submit your solution to the repo using [this solutions guide in our wiki](https://github.com/codeconnector/CodingDojo/wiki#solutions).
+
+*/
+
+function mean(input) {
+	var sum = 0;
+	for (var i = 0; i < input.length; i++) {
+		sum += +input[i][1];
+	};
+	var result = sum / input.length;
+	return +result.toFixed(2);
+};
+
+function median(input) {
+	var numbers = [];
+	for (var i = 0; i < input.length; i++) {
+		numbers.push(+input[i][1]);
+	};
+
+	const sorted = Array.from(numbers).sort((a,b) => a - b);
+	const middle = Math.floor(sorted.length / 2);
+
+	if (sorted.length % 2 === 0) {
+		return (sorted[middle -1] + sorted[middle]) / 2;
+	}
+
+	return sorted[middle];
+};
+
+function higherThan25(input) {
+	let count = 0;
+	for (var i = 0; i < input.length; i++) {
+		if (+input[i][1] > 25) {
+			count++;
+		};
+	};
+	return count;
+};
+
+function twentyFiveToTwentyNine(input) {
+	let count = 0;
+	for (var i = 0; i < input.length; i++) {
+		if (+input[i][1] > 25 && +input[i][1] < 30) {
+			count++;
+		};
+	};
+	return count;
+};
+
+function sortData(inputData, schoolYear, schoolYear2 = 'blah') {
+	let result = inputData.filter(element => element[0] == schoolYear || element[0] == schoolYear2);
+	return result;
+};
+
+function groupData() {
+	// type your solution here
+	fetch('http://justincarter.tech/HS_ACT_Data.txt')
+		.then(response => response.text())
+		.then((responseData) => {
+			const data = responseData.split('\r\n').map((line) => line.split('\t'));
+			//console.log(data);
+
+			let freshmen = sortData(data, 'Freshman');
+			let sophomores = sortData(data, 'Sophomore');
+			let juniors = sortData(data, 'Junior');
+			let seniors = sortData(data, 'Senior');
+			/*
+			console.log('Freshmen: ',freshmen.length,', ',
+				'Sophomores: ',sophomores.length,', ',
+				'Juniors: ',juniors.length,', ',
+				'Seniors: ',seniors.length,', ',
+				'Total: ',data.length);
+			*/
+
+			let sophomoresAndJuniors = sortData(data, 'Sophomore', 'Junior');
+			//console.log('Sophomores and Juniors: ',sophomoresAndJuniors.length);
+
+			console.log(`
```

Comment:
One downside to reporting results via `console.log()` is that you can't _test_ these results. Another is that you can't really integrate such a result into an application (say, if you wanted to present these results to a user of your web site). I returned a `bundle` object above so that it could be tested.

Reply:
Good points, thanks.